### PR TITLE
Implement keystone authentication webhook

### PR DESCRIPTION
**Pull request by alkurbatov**
This pull request brings implementation of keystone authentication webhook.

Please pay attention that build process is clumsy now (one has to enter a lot of data manually). That will be changed in the nearest future.

### CODE REVIEW

**Review by aburluka — APPROVED** [review]

File: Makefile [hunk 1]
```diff
@@ -0,0 +1,11 @@
+BINARY := keystone-auth-hook
+TARGET := build/$(BINARY)
+
+$(TARGET):
+	glide install
```
Comment (ibazhitov) [inline Makefile]: Better add '--strip-vendor' flag to remove nested vendor directories after install.

**Review by ibazhitov — APPROVED** [review]

File: Makefile (hunk 1 above)
Comment (alkurbatov) [reply Makefile]: Thank you, fixed in a separate commit.

File: cmd/main.go [hunk 2]
```diff
@@ -0,0 +1,60 @@
+/*
+Copyright 2017 The Kubernetes Authors.
```
Comment (avagin) [inline cmd/main.go]: Are you sure about "The Kubernetes Authors"?

File: cmd/main.go [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+		fmt.Printf("Usage: %s --certfile <path> --keyfile <path> \n\n", os.Args[0])
+		flag.PrintDefaults()
+	}
+
+	flag.Set("logtostderr", "true")
+	flag.Parse()
+	defer glog.Flush()
+
+	if *certFile == "" || *certKey == "" {
+		flag.Usage()
```
Comment (avagin) [inline cmd/main.go]: It is better to write a reason why it failed

File: cmd/main.go [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+	server := &http.Server{Addr: fmt.Sprintf(":%d", *port)}
+	server.TLSConfig = &tls.Config{
+		// Change default from SSLv3 to TLSv1.0 (because of POODLE vulnerability)
+		MinVersion: tls.VersionTLS10,
+	}
+
+	http.Handle("/authenticate", newAuthWebhook())
+
+	glog.Infof("Serving on :%d", *port)
+	glog.Fatal(server.ListenAndServeTLS(*certFile, *certKey))
```
Comment (avagin) [inline cmd/main.go]: glog.Fatal() always exits with 255, doesn't it?  Does it mean that this script always fail?

I don't like when something useful is executed from log functions. It looks messy. I would suggest to rewrite this code by this way:
```
if err := server.ListenAndServeTLS(*certFile, *certKey): err != nil {
    glog.Fatal(err)
}
```

File: cmd/auth.go [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+func (hook *authWebhook) ServeHTTP(dst http.ResponseWriter, req *http.Request) {
+	if req.Method != http.MethodPost {
+		dst.WriteHeader(http.StatusMethodNotAllowed)
+		return
+	}
+
+	resp := &Payload{}
+	err := json.NewDecoder(req.Body).Decode(resp)
+	if err != nil {
+		glog.Error(err)
```
Comment (avagin) [inline cmd/auth.go]: I think it is better to write more informative messages.
flog.Errorf("Unable to decode a request body: %v", err)

**Comment by avagin** [comment]
Did you think about tests? I can recommend to use travic-ci for basic checks: compilation, golint, etc.

File: cmd/main.go (hunk 2 above)
Comment (alkurbatov) [reply cmd/main.go]: We plan try to merge this webhook to upstream in the nearest future. Until it is done I'll replace the copyrights in a separate commit which could be easily reverted in future.

File: cmd/main.go (hunk 3 above)
Comment (alkurbatov) [reply cmd/main.go]: Will do.

File: cmd/auth.go (hunk 5 above)
Comment (alkurbatov) [reply cmd/auth.go]: Will do.

File: cmd/main.go (hunk 4 above)
Comment (alkurbatov) [reply cmd/main.go]: I've just copy-pasted it from official docs of net/http :) In general http server serves forever unless Ctrl+C is pressed or some fatal error occurred.

**Comment by alkurbatov** [comment]
Yep we need the tests and travis especially for another repo (feathers-authentication-keystone). However there is no travis yet. I need to figure out how it works etc.

I believe the tests will be introduced in a separate pull-request.

**Pull request merged**